File: 0005_Longest_Palindromic_Substring/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        T = "^#" + "#".join(s) + "#$"
        n = len(T)
        P = [0] * n 
        center = right = 0
        
        for i in range(1, n - 1):
            if i < right:
                P[i] = min(right - i, P[2 * center - i])

            while T[i + (1 + P[i])] == T[i - (1 + P[i])]:
                P[i] += 1

            if i + P[i] > right:
                center = i
                right = i + P[i]

        max_len, center_index = max((n, i) for i, n in enumerate(P))
        start = (center_index - max_len) // 2
        return s[start : start + max_len]

File: 0005_Longest_Palindromic_Substring/solution.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    string longestPalindrome(string s) {
        std::string t = "^";
        for (char c : s) {
            t += "#" + std::string(1, c);
        }
        t += "#$";

        int n = t.length();
        vector<int> p(n, 0);
        int center = 0, right = 0;

        for (int i = 1; i < n - 1; i++) {
            int mirror = 2 * center - i;
            
            if (i < right)
                p[i] = min(right - i, p[mirror]);

            while (t[i + (1 + p[i])] == t[i - (1 + p[i])])
                p[i]++;

            if (i + p[i] > right) {
                center = i;
                right = i + p[i];
            }
        }
        int maxLen = 0;
        int centerIndex = 0;
        for (int i = 1; i < n - 1; i++) {
            if (p[i] > maxLen) {
                maxLen = p[i];
                centerIndex = i;
            }
        }

        int start = (centerIndex - maxLen) / 2; 
        return s.substr(start, maxLen);
    }
};